Mathlib/Order/Minimal.lean

```diff
@@ -1,507 +1,645 @@
+  tauto
+
+theorem not_maximal_iff (hx : P x) : ¬ Maximal P x ↔ ∃ y, x < y ∧ P y :=
+  not_minimal_iff (α := αᵒᵈ) hx
+
+theorem exists_of_not_minimal (hx : P x) (h : ¬ Minimal P x) : ∃ y, y < x ∧ P y :=
+  (not_minimal_iff hx).1 h
+
+theorem exists_of_not_maximal (hx : P x) (h : ¬ Maximal P x) : ∃ y, x < y ∧ P y :=
+  (not_maximal_iff hx).1 h
+
+end Preorder
+
+section PartialOrder
+
+variable [PartialOrder α]
+
+theorem Minimal.eq_of_ge (hx : Minimal P x) (hy : P y) (hge : y ≤ x) : x = y :=
+  (hx.2 hy hge).antisymm hge
+
+theorem Minimal.eq_of_le (hx : Minimal P x) (hy : P y) (hle : y ≤ x) : y = x :=
+  (hx.eq_of_ge hy hle).symm
+
+theorem Maximal.eq_of_le (hx : Maximal P x) (hy : P y) (hle : x ≤ y) : x = y :=
+  hle.antisymm <| hx.2 hy hle
+
+theorem Maximal.eq_of_ge (hx : Maximal P x) (hy : P y) (hge : x ≤ y) : y = x :=
+  (hx.eq_of_le hy hge).symm
+
+theorem minimal_iff : Minimal P x ↔ P x ∧ ∀ ⦃y⦄, P y → y ≤ x → x = y :=
+  ⟨fun h ↦ ⟨h.1, fun _ ↦ h.eq_of_ge⟩, fun h ↦ ⟨h.1, fun _ hy hle ↦ (h.2 hy hle).le⟩⟩
+
+theorem maximal_iff : Maximal P x ↔ P x ∧ ∀ ⦃y⦄, P y → x ≤ y → x = y :=
+  minimal_iff (α := αᵒᵈ)
+
+theorem minimal_mem_iff {s : Set α} : Minimal (· ∈ s) x ↔ x ∈ s ∧ ∀ ⦃y⦄, y ∈ s → y ≤ x → x = y :=
+  minimal_iff
+
+theorem maximal_mem_iff {s : Set α} : Maximal (· ∈ s) x ↔ x ∈ s ∧ ∀ ⦃y⦄, y ∈ s → x ≤ y → x = y :=
+  maximal_iff
+
+/-- If `P y` holds, and everything satisfying `P` is above `y`, then `y` is the unique minimal
+element satisfying `P`. -/
+theorem minimal_iff_eq (hy : P y) (hP : ∀ ⦃x⦄, P x → y ≤ x) : Minimal P x ↔ x = y :=
+  ⟨fun h ↦ h.eq_of_ge hy (hP h.prop), by rintro rfl; exact ⟨hy, fun z hz _ ↦ hP hz⟩⟩
+
+/-- If `P y` holds, and everything satisfying `P` is below `y`, then `y` is the unique maximal
+element satisfying `P`. -/
+theorem maximal_iff_eq (hy : P y) (hP : ∀ ⦃x⦄, P x → x ≤ y) : Maximal P x ↔ x = y :=
+  minimal_iff_eq (α := αᵒᵈ) hy hP
+
+@[simp] theorem minimal_ge_iff : Minimal (y ≤ ·) x ↔ x = y :=
+  minimal_iff_eq rfl.le fun _ ↦ id
+
+@[simp] theorem maximal_le_iff : Maximal (· ≤ y) x ↔ x = y :=
+  maximal_iff_eq rfl.le fun _ ↦ id
+
+theorem minimal_iff_minimal_of_imp_of_forall (hPQ : ∀ ⦃x⦄, Q x → P x)
+    (h : ∀ ⦃x⦄, P x → ∃ y, y ≤ x ∧ Q y) : Minimal P x ↔ Minimal Q x := by
+  refine ⟨fun h' ↦ ⟨?_, fun y hy hyx ↦ h'.le_of_le (hPQ hy) hyx⟩,
+    fun h' ↦ ⟨hPQ h'.prop, fun y hy hyx ↦ ?_⟩⟩
+  · obtain ⟨y, hyx, hy⟩ := h h'.prop
+    rwa [((h'.le_of_le (hPQ hy)) hyx).antisymm hyx]
+  obtain ⟨z, hzy, hz⟩ := h hy
+  exact (h'.le_of_le hz (hzy.trans hyx)).trans hzy
+
+theorem maximal_iff_maximal_of_imp_of_forall (hPQ : ∀ ⦃x⦄, Q x → P x)
+    (h : ∀ ⦃x⦄, P x → ∃ y, x ≤ y ∧ Q y) : Maximal P x ↔ Maximal Q x :=
+  minimal_iff_minimal_of_imp_of_forall (α := αᵒᵈ) hPQ h
+
+end PartialOrder
+
+section Subset
+
+variable {P : Set α → Prop} {s t : Set α}
+
+theorem Minimal.eq_of_superset (h : Minimal P s) (ht : P t) (hts : t ⊆ s) : s = t :=
+  h.eq_of_ge ht hts
```

Comment:
It's annoying that those lemmas don't apply to finset

Reply:
I think this is out of scope - nothing about Finset was in the original file.

Reply:
(I've added a TODO)